Control input sequences and amounts

oren-z0 opened the pull request:
Accept one or more sequence values, and one or more input-amount values.
Using repeat_until_length we can make the values the same for all inputs (by giving only values), or different values -
if the length of values is smaller than the number of inputs, values will repeat from start.

Similarly control each output amount, and let it be different from change outputs.

oren-z0 committed:
ignore Mac .DS_Store

oren-z0 committed:
Control input sequences and amounts

Accept one or more sequence values, and one or more input-amount values.
Using repeat_until_length we can make the values the
same for all inputs (by giving only values), or different values -
if the length of values is smaller than the number of inputs,
values will repeat from start.

Similarly control each output amount, and let it be different
from change outputs.

doc-hex commented on psbt_faker/txn.py:
```diff
@@ -81,9 +85,11 @@ def make_change_addr(master_xfp, orig_der,  account_key, idx, style):
 def fake_txn(num_ins, num_outs, master_xpub=None, fee=10000,
          outvals=None, segwit_in=False, wrapped=False, outstyles=None,
-         change_outputs=[], op_return=None, psbt_v2=None, input_amount=1E8,
+         change_outputs=[], op_return=None, psbt_v2=None, input_amount=(1E8,),
```
`input_amount` => `input_amounts`

oren-z0 replied:
Done.

doc-hex commented:
Looks great. Update the README with new output from `--help` and one little rename and I'll merge it.

oren-z0 committed:
update README with new flags

oren-z0 commented:
> Looks great. Update the README with new output from `--help` and one little rename and I'll merge it.

Updated README with new flags. Also changed it a little - instead of deciding the change-amount, it is more intuitive to decide the output-amount(s).

scgbckbone commented on psbt_faker/txn.py:
```diff
@@ -23,6 +23,10 @@ def prandom(count):
     return bytes(random.randint(0, 255) for i in range(count))
 
+def repeat_until_length(values, target_length):
+    # Creates a tuple of target_length length, by repeating the values.
```
this creates a list

scgbckbone replied:
I'd remove `repeat_until_length` from this module, it is not needed here - move to __init__

oren-z0 replied:
Fixed.
Also refactored `fake_txn` and `fake_ms_txn` to accept `input_amounts` as the first parameter - and calculate `num_ins` from it. There is no need to path both of them to the function if `input_amounts` is being extended anyway in:
```
input_amount = repeat_until_length(input_amount, num_ins)
```

scgbckbone replied:
I'd be hesitant to do any changes to `fake_txn` and `fake_ms_txn` because they are just copy/paste from https://github.com/Coldcard/firmware/blob/master/testing/txn.py - so whenever we update it, we will just update here too.

Also if you want, there is more potent version of `fake_{,ms}_txn` in edge https://github.com/Coldcard/firmware/blob/new_edge/testing/txn.py which allows more precise setting of input values.

oren-z0 replied:
@scgbckbone The functions there accept only one `input_amount`.
The new function accepts a list of `input_amounts`. It has to be modified anyways, and it doesn't make sense to accept both `input_amounts` and `num_ins` when the length of `input_amounts` always equals to `num_ins`.

I don't think the purpose of this PR is to standardize the function implementation in all projects.
The way I see it, psbt_faker cli is a standalone tool.

P.S. The cli tool was very useful (for the ultra-paranoids among us) to create fake large transactions and compare the signed results coming from a new air-gapped wallet v.s. an old wallet - making sure there is no [Dark Skippy](https://darkskippy.com). I wanted this transaction to appear as a "normal large transaction" as possible, with different inputs and different outputs.

scgbckbone replied:
> The functions there accept only one input_amount

not true for EDGE, there function accepts `inputs` argument which can be an iterable with up to 4 values: custom subpath, address format aka script type, input amount, and change or not.

anyways, was more like fyi, this is not a big deal and your diff is OK

scgbckbone reviewed (commented):
Some examples in README with new multi-options would be nice
